### ISSUE TITLE
[7.0.0-beta.3] babel-preset-env useBuiltIns "usage" (possibly) adding unnecessary polyfills

### ISSUE THREAD

**Issue by Smolations** [babel/babel#6491]
**Bug report**

I am working with an angular component library developed for an internal application for my organization. It is built via webpack, and therefore uses babel-loader to run transpilation. The current version (not the experimental branch _this_ issue uses) works off v6.26.0 for all babel libs, and v1.6.0 of babel-preset-env. The transpilation works as necessary, but includes _all_ polyfills based on the target environments. In an effort to reduce the size of the final bundle, this experiment was undertaken. Instead of posting a ton of very similar issues for each polyfill problem I encountered (though I will, if desired), I am posting as much as I can in this particular issue. I will try to keep each instance as concise as possible, displaying each issue with only applicable code (as I see it).

#### `es6.function.name` (common, angular-specific)
This issue has to do with the common practice of exporting the name of an angular module while simultaneously defining it. It does not export any function name, but rather the name of the defined module.
```js
// Debug Output
// Added following polyfill:
//  es6.function.name { "android":"4.4.3", "edge":"14", "ie":"11" }

// Input Code (brief)
import angular from 'angular';

export default angular.module('nameThatGetsExported', [])
  .name;
```

#### `es6.array.find` (common, angular-specific)
This issue has to do with the jqLite function `find()` for an angular element. It is library-specific, so no need for a polyfill.
```js
// Debug Output
// Added following polyfill:
//  es6.array.find { "android":"4.4.3", "ie":"11" }

// Input Code (brief)
export default class SomethingElementService {
  /* @ngInject */
  constructor(/* $compile, other angular injectables */) {
    // e.g. many of these...
    // this.$compile = $compile;
  }

  makeSomething(options = {}) {
    // this._options.parent is expected to be a jqLite element
    // this is the _only_ reference to a `find()` method in this file
    const $something = this.$element || this._options.parent.find('something-element');
    // ...
  }
}
```

#### `es6.object.set-prototype-of`, `es6.symbol`
This polyfill was only included in files that define angular controllers that extend a base angular controller class. This may not be an issue, but I thought there was already a plugin that handled class transpilations (e.g. `transform-es2015-classes`).
```js
// Debug Output
// Added following polyfill:
//  es6.object.set-prototype-of { "android":"4.4.3" }
//  es6.symbol { "android":"4.4.3", "edge":"14", "ie":"11" }

// Input Code (only names of controllers changed, everything else as-is)
import BaseCtrl from './base-ctrl.es6';

export default class ExtendedCtrl extends BaseCtrl {
  /* @ngInject */
  constructor($element, $transclude) {
    super($element, $transclude);
  }
}
```

#### `es6.regexp.match`, `es6.regexp.match`
This issue is similar to the issue with `es6.array.find` in that it seems to just see the method names and then add the polyfills. There may be confusion with the string methods that have the same name and are using RegExps for the pattern. It should be noted that some of the code in the example is working with jqLite elements, and lodash is being used. Also, for brevity, a similar polyfill, `es6.regexp.split`, is added in another file with a similar implementation as below, where `.split()` is being called on a string variable using a RegExp as the pattern.
```js
// Debug Output
// Added following polyfill:
//  es6.regexp.match { "android":"4.4.3", "edge":"14", "ie":"11" }
//  es6.regexp.search { "android":"4.4.3", "edge":"14", "ie":"11" }

// Input Code (brief; only relevant code with match/search methods shown)
import _ from 'lodash';

export default class SomeCtrl {
  /* @ngInject */
  constructor(/*angular injectables*/) {
    // many of these...
    // this.$injectable = $injectable;
  }

  _handleKeyup(e) {
    const rawInput = this.inputElement.val();
    const before = rawInput.substring(0, this.inputElement[0].selectionStart).match(/[@\w]*$/)[0];
    const after = rawInput.substring(this.inputElement[0].selectionStart, rawInput.length).match(/^[@\w]*/)[0];
    const currentWord = before + after;
    if (_.includes(currentWord, '@')) {
      this._search(currentWord);
    } else {
      this._closeMenu();
    }
  }

  _handleSelected(user) {
    const rawInput = this.inputElement.val();
    const selectionStart = this.inputElement[0].selectionStart;
    const left = rawInput.slice(0, selectionStart);
    const leftPos = left.search(/\S+$/);
    // ...
  }

  _search(query) {
    return this.someService.search(_.trimStart(query, '@'), this.contextType, this.contextId)
      .then(/*...*/)
      // ...
  }
}
```

### Babel Configuration (.babelrc, package.json, cli command)
```js
{
  "presets": [
    ["env", {
      "modules": false,
      "targets": {
        // browserlist combined by OR clause
        // https://github.com/ai/browserslist
        "browsers": ["last 2 versions", "not ie <= 10"],
      },
      "useBuiltIns": "usage",
      "debug": true,
    }],
  ],
  "plugins": [
    [
      "transform-es2015-modules-commonjs-simple", {
        "noMangle": true, // needed or DI explodes
      },
    ],
  ],
}
```

### Expected Behavior
No polyfills are needed/added.

### Current Behavior
Polyfills added.

### Possible Solution
It seems like babel has some trouble determining when some functions are defined by a third party library or by an implementor, and instead just looks for some matches based on name alone. I realize that this is incredibly difficult to trace, especially if any directories are excluded from transpilation. Perhaps there should be some sort of comment exclusion (like eslint uses with `/* eslint-disable */` comments) to explicitly tell babel not to include some polyfills in situations where it may not always be realistic for babel-preset-env/babel to determine polyfill necessity itself.

### Context
Adding unnecessary polyfills increases the overall bundle size. While it appears that there are less overall polyfills being added than the previous version of babel-preset-env, I imagine that striving for accuracy with the new functionality is desirable.

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel(core, preset-env, polyfill)            |  7.0.0-beta.3
| webpack      | 3.5.5
| babel-loader    | 7.1.2
| node             | 6.10.3
| npm              | 3.10.10
| Operating System | Ubuntu 16.04 (Vagrant)

**Comment by hzoo** [comment]
@Smolations yeah this is expected atm, I figured it would make sense to include in case which is better than including too many polyfills (it is already better than before). We can only fix this by type info really, or some kind of heuristic

The workaround is that you can use the `exclude` option in the meantime.

**Comment by Smolations** [comment]
@hzoo So does that mean this will be the expected result for the initial 7.x release and (maybe) addressed in a later release?

**Comment by hzoo** [comment]
It's the expected result unless I/you/someone else has time to look into how to make it better. I don't really call it a bug because there is no way to be sure you should/shouldn't add a polyfill. I have a lot more discussion/thoughts in https://github.com/babel/babel-preset-env/issues/284 so I would read that

**Comment by Smolations** [comment]
@hzoo Alright, so should I close this issue then or leave it open for consideration later?

**Comment by hzoo** [comment]
We can def leave it upon and I certainly want to make it better, I'm just saying it's difficult to do

Maybe can do something like your comment idea as a workaround as well

**Comment by Smolations** [comment]
Alright, I will leave it open and contribute to the conversation in babel/babel-preset-env#284. Thanks! 😋 👍

**Comment by Smolations** [comment]
Also, just putting a bird in your ear as it's not worth opening a whole other issue for, but could there be a `'verbose'` choice or something for the `debug` option with the babel-preset-env? I don't need to know the entire list of files that _don't_ get polyfills. I'm only interested in the ones that do.  😇

**Comment by ylemkimon** [comment]
@hzoo Is function.name being polyfilled a right behavior even when, name is used like `object[name]` (name is a variable, not a string literal)?

**Comment by LeuisKen** [comment]
> Since this would be a file-by-file issue, I would propose some sort of comment-disabling system, similar to eslint's /* eslint-disable */ system.

#6626

I have just tried this solution and write a small demo, just add this two functions in the https://github.com/babel/babel/blob/master/packages/babel-preset-env/src/use-built-ins-plugin.js

```js
/**
 * check if developer disabled polyfill
 * 
 * @param {NodePath} path the checking node path
 * @return {boolean} if false, it means that someone has written
 *    `// # polyfill disabled`
 */
function shouldUseBuiltIns(path) {
  let parentPath = path.parentPath;
  while (parentPath) {
    let leadingComments = parentPath.node.leadingComments;
    if (Array.isArray(leadingComments)) {
      const activeComments = leadingComments
        .filter(comment => comment.type === 'CommentLine')
        .filter(comment => hasSubString(comment.value, '# polyfill'));
      if (activeComments.length > 0) {
        let commentStr = activeComments.pop().value;
        if (hasSubString(commentStr, 'disabled')) {
          return false;
        }
        return true;
      }
    }
    parentPath = parentPath.parentPath;
  }
  return true;
}

function hasSubString(string, subString) {
  return string.indexOf(subString) !== -1;
}
```

and call `shouldUseBuiltIns` in the `addImport ` function: https://github.com/babel/babel/blob/master/packages/babel-preset-env/src/use-built-ins-plugin.js#L43

Then we have:

**Sample input:**

```js
// # polyfill disabled
const guidPrefix = (new Date()).getTime().toString(16).slice(8);
// # polyfill enabled
```

**Sample output:**

```js
// # polyfill disabled
var guidPrefix = new Date().getTime().toString(16).slice(8); // # polyfill enabled
```

which will not `import "core-js/modules/es6.regexp.to-string";` anymore.

Since this issue is still in discussion, I will not make a pull request until we have an conclusion of this issue.
